syntax = "proto3";

package straxrpc;

service StraxRPC {

  rpc SearchField(SearchPattern) returns (stream ColumnInfo) {}

  rpc DataInfo(PluginInfo) returns (stream DataColumn) {}

  rpc GetDF(TableInfo) returns (stream DataColumn) {}

  rpc GetArray(TableInfo) returns (stream DataColumn) {}
}

message SearchPattern {
    string pattern = 1;
    uint32 max_matches = 2;

}

message TableInfo {
  string name = 1;
  string run_id = 2;

}

message PluginInfo {
  string name = 1;
  string class_name = 2;
  repeated ColumnInfo columns = 3;
}


message ColumnInfo {
  string name  = 1;
  string data_name = 2;
  string plugin = 3;
  string comment = 4;
  string dtype = 5;

}


message Int32Array {
  repeated int32 values = 1;
}

message Int64Array {
  repeated int64 values = 1;
}

message Int32ArrayArray {
  repeated Int32Array values = 1;
}

message Int64ArrayArray {
  repeated Int64Array values = 1;
}

message BoolArray {
  repeated bool values = 1;
}

message Float32Array {
  repeated float values = 1;
}

message Float64Array {
  repeated double values = 1;
}

message Float32ArrayArray {
  repeated Float32Array values = 1;
}

message Float64ArrayArray {
  repeated Float64Array values = 1;
}

message StringArray {
  repeated string values = 1;
}



message DataColumn {
  ColumnInfo info = 99;
  repeated uint32 index = 1;
  oneof data {
    Int32Array int32 = 2;
    Int64Array int64 = 3;
    Float32Array float32 = 4;
    Float64Array float64 = 5;
    StringArray string = 6;
    BoolArray bool = 7;
    StringArray object = 8;
    Float32ArrayArray float32array = 9;
    Float64ArrayArray float64array = 10;
    Int32ArrayArray int32array = 11;
    Int64ArrayArray int64array = 12;
  }
}

